[2025-07] Shopify major CalVer release enforcement

juanpprieto opened the pull request:
## CalVer Release Enforcement Automation

### Why is this PR introduced?

Hydrogen uses a sophisticated release system that combines:
- **Changesets** for managing package versions and changelogs
- **CalVer (Calendar Versioning)** for core packages aligned with Shopify's quarterly API releases
- **Semver** for CLI and tooling packages

Previously, CalVer enforcement was a **manual process** that required:
- Manually adjusting versions after changesets ran
- Remembering to align major versions to quarters (Q1=1, Q2=4, Q3=7, Q4=10)
- Manually updating CHANGELOG headers to match CalVer format
- Risk of human error in version coordination

### What this PR does

This PR **introduces automation** for our CalVer enforcement by:

1. **Creating two new automation scripts:**
   - `enforce-calver-ci.js` (186 lines) - Production script that runs automatically in CI/CD
   - `enforce-calver-local.js` (273 lines) - Development tool with dry-run and safety features

2. **Adding comprehensive documentation** (`docs/CALVER.md`) that explains:
   - How CalVer integrates with changesets
   - The automated two-step versioning process
   - Quarter alignment for major releases
   - Package coordination strategies

3. **Creating robust testing infrastructure:**
   - `test-calver-comparison.sh` - Compares standard vs CalVer versioning
   - GitHub Actions workflow for Linux CI validation
   - Cross-platform compatibility (macOS and Linux)

###  How It Works

The new CalVer automation leverages changesets' features while adding CalVer transformation:

```
1. Changesets runs → Applies semver bumps (e.g., 2025.5.0 → 2025.5.1)
2. CalVer script runs → Automatically transforms to CalVer format
3. Post-processing → Updates CHANGELOGs, dependencies, and version files
```

**Automated behaviors:**
- **Patch/Minor bumps**: Increment third segment (2025.5.0 → 2025.5.1)
- **Major bumps**: Automatically align to next quarter (2025.5.0 → 2025.7.0 for Q3)
- **Mixed packages**: CalVer for hydrogen/hydrogen-react, semver for CLI tools

###  Testing/Tophatting

The PR includes:
- Automated CI tests that validate script behavior on Linux
- Comparison tests showing before/after versioning
- Dry-run mode for safe local testing
- Cross-platform sed compatibility

### Impact

This automation:
- **Eliminates manual work** - No more manual version adjustments after changesets
- **Prevents errors** - Automatic quarter alignment and CHANGELOG updates
- **Saves time** - What was manual is now automatic
- **Ensures consistency** - Same rules applied every time, no human variance
- **Enables confidence** - CI tests ensure the automation works correctly

### Related Documentation

See \`docs/CALVER.md\` for detailed information about:
- How CalVer packages are coordinated automatically
- Quarterly release alignment with Storefront API
- Testing and validation procedures
- Usage in both CI and local development

---

**Reviewer Notes:**
- This PR introduces automation for what was previously a manual process
- All feedback from @kdaviduik and GitHub Copilot has been addressed
- Tests are non-destructive and safe to run
- The comparison script helps visualize version transformations

juanpprieto committed:
calver enforcing wip

juanpprieto committed:
refactor: Split CalVer enforcement into CI and local scripts

BREAKING: Replaces single 740-line enforce-calver-versioning.js with:
- enforce-calver-ci.js (177 lines) - lean production-only script
- enforce-calver-local.js (250 lines) - local testing with dry-run

Changes:
- 76% LOC reduction for production script
- Clear separation between CI and local development workflows
- CI script runs without safeguards (designed for releases)
- Local script defaults to dry-run (requires --apply flag)
- Removed redundant helper functions and test code
- Added comprehensive CALVER.md documentation
- Added test:calver npm script for comparison testing

The CalVer enforcement continues to:
- Transform versions to YYYY.MAJOR.MINOR format
- Align major versions to quarters (1,4,7,10)
- Coordinate linked packages on major bumps only
- Validate against version regression

juanpprieto committed:
refactor: streamline CalVer enforcement with split CI/local scripts

- Reduce CalVer script from 740 to 177 lines (76% reduction)
- Split into enforce-calver-ci.js (production) and enforce-calver-local.js (testing)
- Move scripts to .changeset/ directory for better organization
- Add comprehensive documentation in docs/CALVER.md
- Update package.json to use new CI script path
- Remove old monolithic script and test files
- Maintain all CalVer functionality with cleaner separation of concerns

kdaviduik commented on docs/CALVER.md:
```diff
@@ -0,0 +1,161 @@
+- **YYYY**: Current year
+- **MAJOR**: Quarter-aligned (1, 4, 7, 10) for Q1, Q2, Q3, Q4
+- **MINOR**: Feature releases within a quarter
+- **PATCH**: Bug fixes (adds 4th segment: YYYY.MAJOR.MINOR.PATCH)
```
This isn't correct - we don't differentiate between patches and minor changes in our calver system

kdaviduik replied:
We have accurate docs on this [here](https://github.com/Shopify/hydrogen/blob/main/CLAUDE.md#version-strategy)

juanpprieto committed:
fix: address PR feedback from kdaviduik and Copilot

- Remove incorrect patch/minor differentiation in CalVer docs (kdaviduik)
- Fix duplicate changeset version execution (Copilot suggestion #1)
- Fix CHANGELOG header replacement to use correct version (Copilot suggestion #2)
- Add clarifying comment about the semver-to-CalVer transformation flow

juanpprieto committed:
fix: make sed commands portable for Linux CI environments

- Add detection for GNU sed (Linux) vs BSD sed (macOS)
- Use appropriate sed -i syntax for each platform
- Ensures scripts work on GitHub Actions (Linux) and local development (macOS)
- Addresses Copilot feedback about cross-platform compatibility

juanpprieto committed:
test: add CI workflow for CalVer validation (Linux only)

- Focus on CI-critical components only
- Test 1: CI script syntax validation
- Test 2: GNU sed Linux compatibility
- Test 3: Patch version bumps (2025.5.0 → 2025.5.1)
- Test 4: Major bumps with quarter alignment (2025.5.0 → 2025.7.0)
- Test 5: Mixed CalVer/semver package handling
- All tests are safe and non-destructive
- Removed non-essential local script testing
